src/Compilers/Core/AnalyzerDriver/AnalyzerExecutor.cs

```diff
@@ -466,7 +476,18 @@ internal static bool CanHaveExecutableCodeBlock(ISymbol symbol)
 
         internal void ExecuteAndCatchIfThrows(DiagnosticAnalyzer analyzer, Action analyze)
         {
-            ExecuteAndCatchIfThrows(analyzer, analyze, _onAnalyzerException, _cancellationToken);
+            object gate;
+            if (_singleThreadedAnalyzerToGateMap.TryGetValue(analyzer, out gate))
+            {
+                lock(gate)
```

Comment:
actual driver seems async, can we use Async Semaphore here rather than lock and make this one async as well? that way we probably won't waste a thread due to a lock.